### PR TITLE
Radius Outlier Filter for Tensor PointCloud

### PR DESCRIPTION
**Pull request by reyanshsolis**
Python Example:
```python
import open3d as o3d

# Sample PCD data.
sample_pcd_data = o3d.data.PCDPointCloud()

# Load the data, and down-sample.
tpcd = o3d.t.io.read_point_cloud(sample_pcd_data.path).voxel_down_sample(0.02)
o3d.visualization.draw(tpcd)

# Remove Radius Outliers. 
tpcd_filtered, indices = tpcd.remove_radius_outliers(nb_points=20, search_radius=0.05)
o3d.visualization.draw(tpcd_filtered)

# Print len of points before and after to check.
print(len(tpcd.point["positions"]))
print(len(tpcd_filtered.point["positions"]))
```

Performance Comparision (for `o3d.data.PLYPointCloud`)
```
---------------------------------------------------------------------------------------
Benchmark                                             Time             CPU 
---------------------------------------------------------------------------------------
RemoveRadiusOutliers/CPU[50 | 0.05]                46.5 ms         46.0 ms 
RemoveRadiusOutliers/CUDA[50 | 0.05]               2.05 ms         2.04 ms
LegacyRemoveRadiusOutliers/Legacy[50 | 0.05]       59.2 ms         57.9 ms
```

TODO:
- [ ] Add Unit Tests.
- [x] Merge after https://github.com/isl-org/Open3D/pull/4955 is merged to master.

This change is [Reviewable](https://reviewable.io/reviews/isl-org/open3d/4959)

### CODE REVIEW

**Comment by Tarush-Singh35** [comment]
@reyanshsolis i even i want to learn 3D programming and want to contribute in the Projects tell me how and where can i start

File: cpp/open3d/t/geometry/PointCloud.h [hunk 1]
```diff
@@ -311,12 +311,49 @@ class PointCloud : public Geometry, public DrawableGeometry {
...
+    ///
+    /// Points with indices in \p indices are selected.
+    ///
+    /// \param indices Boolean indexing tensor of shape {n,} containing
+    /// true value for the indices that is to be selected.
+    /// \param invert Set to `True` to invert the selection of indices.
+    PointCloud SelectByIndex(const core::Tensor &indices,
+                             bool invert = false) const;
+
+    /// \brief Returns `True` is two pointcloud are similar.
```
Comment (nigels-com) [inline cpp/open3d/t/geometry/PointCloud.h:325]: **_if_** two pointcloud are similar.

**Review by yxlao — CHANGES_REQUESTED** [review]
> Reviewed 1 of 6 files at r1, 5 of 7 files at r2, 4 of 4 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/4959)* status: all files reviewed, 14 unresolved discussions (waiting on @nigels-com and @reyanshsolis)

___
*[cpp/open3d/t/geometry/PointCloud.h, line 325 at r3](https://reviewable.io/reviews/isl-org/open3d/4959#-MzVgnSJez6nrHCOBVLS-r3-325:-N-8X0jxCx5bqqm2_IRz:byahial) ([raw file](https://github.com/isl-org/open3d/blob/53f7717fd987de8dea62b5516e657e5515df9338/cpp/open3d/t/geometry/PointCloud.h#L325)):*
<details><summary><i>Previously, nigels-com wrote…</i></summary><blockquote>

**_if_** two pointcloud are similar.
</blockquote></details>

point cloud -> point clouds
___
*[cpp/open3d/t/geometry/PointCloud.h, line 327 at r4](https://reviewable.io/reviews/isl-org/open3d/4959#-N-8ZaC79aa1M3nCZIg6:-N-8ZaC79aa1M3nCZIg7:batqyq6) ([raw file](https://github.com/isl-org/open3d/blob/4aaec71463613b94cc02539dcda3e5f89d6cb2fd/cpp/open3d/t/geometry/PointCloud.h#L327)):*
> ```objc
>     /// \brief Returns `True` is two pointcloud are similar.
>     ///
>     /// Two unorganised pointclouds are similar when, between the
> ```

nit: unorganized
___
*[cpp/open3d/t/geometry/PointCloud.h, line 328 at r4](https://reviewable.io/reviews/isl-org/open3d/4959#-N-8XkfT0iGFH0seu5Z1:-N-8XkfT0iGFH0seu5Z2:bax8q3c) ([raw file](https://github.com/isl-org/open3d/blob/4aaec71463613b94cc02539dcda3e5f89d6cb2fd/cpp/open3d/t/geometry/PointCloud.h#L328)):*
> ```objc
>     ///
>     /// Two unorganised pointclouds are similar when, between the
>     /// correspondences of the two pointcloud the inliear fitness
> ```

nit: point cloud needs to be plural. Check plural forms in this PR
___
*[cpp/open3d/t/geometry/PointCloud.h, line 336 at r4](https://reviewable.io/reviews/isl-org/open3d/4959#-N-8X9P93loe6Clkm0Aj:-N-8X9P93loe6Clkm0Ak:b-e1mf16) ([raw file](https://github.com/isl-org/open3d/blob/4aaec71463613b94cc02539dcda3e5f89d6cb2fd/cpp/open3d/t/geometry/PointCloud.h#L336)):*
> ```objc
>     /// \param inlier_rmse_threshold the inliear_rmse must be smaller than this
>     /// threshold to be similar.
>     bool IsSimilar(const PointCloud &other,
> ```

`PointCloud::IsSimilar` may not be general enough to be a top-level `PointCloud` function. I suggest moving it to to the `RemoveRadiusOutliers` function directly.  In fact the implementation is just a few lines, so it makes more sense to integrate into `RemoveRadiusOutliers`.
___
*[cpp/open3d/t/geometry/PointCloud.h, line 347 at r4](https://reviewable.io/reviews/isl-org/open3d/4959#-N-8Xb1Y9skhIEiHQP58:-N-8Xb1Y9skhIEiHQP59:b-tt38mf) ([raw file](https://github.com/isl-org/open3d/blob/4aaec71463613b94cc02539dcda3e5f89d6cb2fd/cpp/open3d/t/geometry/PointCloud.h#L347)):*
> ```objc
>                                        core::HashBackendType::Default) const;
> 
>     /// \brief Function to remove points that have less than \p nb_points in a
> ```

- Remove `Function to`. Same for other docstrings.
- `\p nb_points` -> `\p nb_points neighbors`
___
*[cpp/open3d/t/geometry/PointCloud.h, line 350 at r4](https://reviewable.io/reviews/isl-org/open3d/4959#-N-8Xx-b9mIpCiLp0r-1:-N-8Xx-b9mIpCiLp0r-2:b-9vebig) ([raw file](https://github.com/isl-org/open3d/blob/4aaec71463613b94cc02539dcda3e5f89d6cb2fd/cpp/open3d/t/geometry/PointCloud.h#L350)):*
> ```objc
>     /// sphere of a given radius.
>     ///
>     /// \param nb_points Number of neighbour points required within the radius.
> ```

nit: neighbor (consistent with the neighbor search function)
___
*[cpp/open3d/t/geometry/PointCloud.h, line 352 at r4](https://reviewable.io/reviews/isl-org/open3d/4959#-N-8YzX64-ssrXRBs1gS:-N-8YzX64-ssrXRBs1gT:b-b693s3) ([raw file](https://github.com/isl-org/open3d/blob/4aaec71463613b94cc02539dcda3e5f89d6cb2fd/cpp/open3d/t/geometry/PointCloud.h#L352)):*
> ```objc
>     /// \param nb_points Number of neighbour points required within the radius.
>     /// \param search_radius Radius of the sphere.
>     /// \return tuple of filtered PointCloud and boolean indexing tensor
> ```

It is either a "boolean mask tensor", or an integer "indexing tensor". Indexing tensor cannot be boolean. 
___
*[cpp/open3d/t/geometry/PointCloud.h, line 355 at r4](https://reviewable.io/reviews/isl-org/open3d/4959#-N-8ZJsg8drlXQWsQpHZ:-N-8ZJsg8drlXQWsQpH_:b-ccq7wk) ([raw file](https://github.com/isl-org/open3d/blob/4aaec71463613b94cc02539dcda3e5f89d6cb2fd/cpp/open3d/t/geometry/PointCloud.h#L355)):*
> ```objc
>     /// w.r.t. input pointcloud.
>     std::tuple<PointCloud, core::Tensor> RemoveRadiusOutliers(
>             size_t nb_points, double search_radius) const;
> ```

`nb_points` -> ``
___
*[cpp/open3d/t/geometry/PointCloud.cpp, line 220 at r4](https://reviewable.io/reviews/isl-org/open3d/4959#-N-8YT9M9Vp0Sb3-ZYl9:-N-8YT9M9Vp0Sb3-ZYlA:b-wmcvam) ([raw file](https://github.com/isl-org/open3d/blob/4aaec71463613b94cc02539dcda3e5f89d6cb2fd/cpp/open3d/t/geometry/PointCloud.cpp#L220)):*
> ```cpp
>                                      bool invert /* = false */) const {
>     const int64_t length = this->GetPointPositions().GetLength();
>     core::AssertTensorDtype(indices, core::Dtype::Bool);
> ```

The name is not consistent.  We say `SelectByIndex` but we are actually doing "select by boolean mask". We should change to actually select by index.
___
*[cpp/open3d/t/geometry/PointCloud.cpp, line 256 at r4](https://reviewable.io/reviews/isl-org/open3d/4959#-N-8_Ovz7N5YTcG8qmAg:-N-8_Ovz7N5YTcG8qmAh:bsrygz0) ([raw file](https://github.com/isl-org/open3d/blob/4aaec71463613b94cc02539dcda3e5f89d6cb2fd/cpp/open3d/t/geometry/PointCloud.cpp#L256)):*
> ```cpp
>                       result.inlier_rmse_);
> 
>     if (result.fitness_ >= inlier_fitness_threshold &&
> ```

do it without if-else
___
*[cpp/open3d/t/geometry/PointCloud.cpp, line 312 at r4](https://reviewable.io/reviews/isl-org/open3d/4959#-N-8YqZo6Z9H9aWnHrnL:-N-8YqZo6Z9H9aWnHrnM:bm9ph2a) ([raw file](https://github.com/isl-org/open3d/blob/4aaec71463613b94cc02539dcda3e5f89d6cb2fd/cpp/open3d/t/geometry/PointCloud.cpp#L312)):*
> ```cpp
> 
>     int64_t size = row_splits.GetLength();
>     core::Tensor num_neighbours =
> ```

nit: US spelling
___
*[cpp/open3d/t/geometry/PointCloud.cpp, line 315 at r4](https://reviewable.io/reviews/isl-org/open3d/4959#-N-8Zn2b9zBZXxtouG5p:-N-8Zn2b9zBZXxtouG5q:bmqupkg) ([raw file](https://github.com/isl-org/open3d/blob/4aaec71463613b94cc02539dcda3e5f89d6cb2fd/cpp/open3d/t/geometry/PointCloud.cpp#L315)):*
> ```cpp
>             row_splits.Slice(0, 1, size) - row_splits.Slice(0, 0, size - 1);
> 
>     core::Tensor valid = num_neighbours.Ge(static_cast<int64_t>(nb_points));
> ```

In the docstring we say "Function to remove points that have **less than** \p nb_points", but this is `Ge`. To be equivalent, we should use greater-than-or-equal-to. 



_Code quote (from cpp/open3d/t/geometry/PointCloud.h):_
```objc
 Function to remove points that have less than \p nb_points 
```
___
*[cpp/pybind/t/geometry/pointcloud.cpp, line 185 at r4](https://reviewable.io/reviews/isl-org/open3d/4959#-N-8_pri3R58-EACNqHw:-N-8_pri3R58-EACNqHx:b-s9eziv) ([raw file](https://github.com/isl-org/open3d/blob/4aaec71463613b94cc02539dcda3e5f89d6cb2fd/cpp/pybind/t/geometry/pointcloud.cpp#L185)):*
> ```cpp
> 
>     pointcloud.def("select_by_index", &PointCloud::SelectByIndex, "indices"_a,
>                    "invert"_a = false);
> ```

Inject docs.
___
*[cpp/pybind/t/geometry/pointcloud.cpp, line 207 at r4](https://reviewable.io/reviews/isl-org/open3d/4959#-N-8_nrKEIsXPRZFL9ZZ:-N-8_nrKEIsXPRZFL9Z_:b-s9eziv) ([raw file](https://github.com/isl-org/open3d/blob/4aaec71463613b94cc02539dcda3e5f89d6cb2fd/cpp/pybind/t/geometry/pointcloud.cpp#L207)):*
> ```cpp
>             "nb_points"_a, "search_radius"_a,
>             "Function to remove points that have less than nb_points in a "
>             "sphere of a given radius.");
> ```

Inject docs.

**Comment by reyanshsolis** [comment]
@yxlao moved to #4990.

**Pull request closed without merging**